File: Java/TreeisBSTorNot.java
//Given a binary tree check if it is binary search tree or not.


import java.util.Scanner;

 class Result {
    static int a[]=new int[10];
    static int i=0;
    static void inorder(Node p)
    {
      if(p!=null)
      {
        inorder(p.leftChild);
        a[i++]=p.data;
        inorder(p.rightChild);
      }
    }
    static int isBinarySearchTree(Node t1) {
      inorder(t1);
      for(int j=0;j<i-1;j++)
      {
        if(a[j]>a[j+1])
          return 0;
          
      }
      return 1;
    }
  }


class Node {
    int data; // data used as key value
    Node leftChild;
    Node rightChild;
    public Node()  {
      data=0;
    }
    public Node(int d)  {
      data=d;
    }
  }
  
  class Main
  {
    static void inOrder(Node root)
    {
      if (root != null)
      {
        inOrder(root.leftChild);
        System.out.print(root.data + " ");
        inOrder(root.rightChild);
      }
    }
    static Node insertLevelOrder(int arr[], Node root, int i, int n)
    {
      if (i < n)
      {
        Node temp = new Node(arr[i]);
        root = temp;
        root.leftChild = insertLevelOrder(arr, root.leftChild, 2 * i + 1, n);
        root.rightChild = insertLevelOrder(arr, root.rightChild, 2 * i + 2, n);
      }
      return root;
    }
  
    static Node buildTree(int t[], int n)
    {
      Node node = null;
      node = insertLevelOrder(t, node, 0, n);
      return node;
    }
    public static void main(String[] args)
    {
      int a[]=new int[10];
      int n, m, i;
      Scanner s=new Scanner(System.in);
      n = Integer.parseInt(s.nextLine());
      for(i=0;i<n;i++)
        a[i] = s.nextInt();
      Node root = null;
      if(n!=0)
        root = buildTree(a, n);
      System.out.println(Result.isBinarySearchTree(root));
    }
  }

File: C/TreeisBSTorNot.c
//Given a binary tree check if it is binary search tree or not.

#include<stdio.h>
#include<stdlib.h>

struct Node
{
  int data;
  struct Node *left, *right;
};

struct Node* newNode(int data)
{
  struct Node* node = (struct Node*)malloc(sizeof(struct Node));
  node->data = data;
  node->left = node->right = NULL;
  return (node);
}

struct Node* insertLevelOrder(int arr[], struct Node* root, int i, int n)
{
  if (i < n)
  {
    struct Node* temp = newNode(arr[i]);
    root = temp;

    root->left = insertLevelOrder(arr, root->left, 2 * i + 1, n);

    root->right = insertLevelOrder(arr, root->right, 2 * i + 2, n);
  }
  return root;
}

struct Node* buildTree(int t[], int n)
{
  struct Node* root = (struct Node*)malloc(sizeof(struct Node));
  root = insertLevelOrder(t, root, 0, n);
  return(root);
}


static int a[10],i=0;

void inorder(struct Node* p)
{
  if(p!=NULL)
  {
    inorder(p->left);
    a[i++]=p->data;
    inorder(p->right);
  }
}

int isBinarySearchTree(struct Node* t1)
{
  inorder(t1);
  for(int j=0;j<i-1;j++)
  {
    if(a[j]>a[j+1])
    {
      return 0;
    }
  }
  return 1;

}




int main()
{
  int a[10], n, i;
  scanf("%d", &n);
  for(i=0;i<n;i++)
    scanf("%d", &a[i]);
  struct Node* root = NULL;
  root = buildTree(a, n);
  printf("%d", isBinarySearchTree(root));
  return 0;
}
